### PR TITLE
Create new codec type for Compact

### PR DESCRIPTION
**Pull request by gabrielokura**
## Description
This PR adds a new codec type for encode/decode Compact integers.

## Why?
Refactoring `Codec`. 

## How?
1. Create new CodecCompact class
2. Create its tests for encoding/decoding values

## Testing Plan 
Run `dart test` in polkadart_scale_codec package folder

### CODE REVIEW

**Comment by gabrielokura** [comment]
@pastre If you could check this out. I think your insights would be much helpful in this different approach

File: packages/polkadart_scale_codec/lib/src/types/codec_compact.dart [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  /// ```
+  /// final intEncoded = CodecCompact<int>().encodeToHex(65535); //"0xfeff0300"
+  /// final bigIntEncoded = CodecCompact<BigInt>().encodeToHex(18446744073709551615.toBigInt);
+  /// //0x13ffffffffffffffff
+  /// ```
+  @override
+  String encodeToHex(value) {
+    var sink = HexEncoder();
+
+    var convertedValue = value is int ? value : (value as BigInt).toInt();
```
Comment (pastre) [inline packages/polkadart_scale_codec/lib/src/types/codec_compact.dart:47]: `(value as BigInt).toInt()` might be a bad idea, as values larger than what `int` supports will be clamped Look for an approach where you don't need to convert types, nor check for them in real time. A simple approach would be to treat everything as a bigint

**Review by pastre — CHANGES_REQUESTED** [review]

**Comment by pastre** [comment]
Generics is a tool that allows us to not specify types while keeping all the benefits from string typing. If we're gonna type cast and throw, it might be simpler to just keep stuff as simple as possible

File: packages/polkadart_scale_codec/lib/src/types/codec_compact.dart (hunk 1 above)
Comment (gabrielokura) [reply packages/polkadart_scale_codec/lib/src/types/codec_compact.dart]: Done! Changed CodecCompact to support only BigInt and let the user manipulates the result and transform to int (if wants)

File: packages/polkadart_scale_codec/lib/src/types/codec_compact.dart [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  /// Example:
+  /// ```
+  /// String example1 = CodecCompact().encodeToHex('65535'.toBigInt); //"0xfeff0300"
+  /// String example2 = CodecCompact().encodeToHex(18446744073709551615.toBigInt);
+  /// //0x13ffffffffffffffff
+  /// ```
+  @override
+  String encodeToHex(value) {
+    var sink = HexEncoder();
+    int convertedValue = value.toInt();
```
Comment (pastre) [inline packages/polkadart_scale_codec/lib/src/types/codec_compact.dart:37]: Conversion to int might be clamped. Why aren't we using bigint?

Comment (pastre) [reply packages/polkadart_scale_codec/lib/src/types/codec_compact.dart:37]: If you need it to be an int, consider creating a guard clause for using  `value.isValidInt`. This will mitigate the pyramid of doom that's happening down here

File: packages/polkadart_scale_codec/lib/src/types/codec_compact.dart [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    if (convertedValue < 64) {
+      sink.write(convertedValue * 4);
+    } else if (convertedValue < pow(2, 14).toInt()) {
+      sink.write((convertedValue & 63) * 4 + 1);
+      sink.write(convertedValue >>> 6);
+    } else if (convertedValue < pow(2, 30).toInt()) {
+      sink.write((convertedValue & 63) * 4 + 2);
+      sink.write((convertedValue >>> 6) & 0xff);
+      sink._uncheckedU16(convertedValue >>> 14);
+    } else if (convertedValue < calculateBigIntPow(2, 536).toInt()) {
```
Comment (pastre) [inline packages/polkadart_scale_codec/lib/src/types/codec_compact.dart:52]: `calculateBigIntPow(2, 536)` doesn't fit in an int (dart ints are as large as 2^53) and is getting clamped to 2^53. Really unsure if this is working

File: packages/polkadart_scale_codec/test/test_types/codec_compact_test.dart [hunk 4]
```diff
@@ -0,0 +1,104 @@
+import 'package:polkadart_scale_codec/polkadart_scale_codec.dart';
+import 'package:test/test.dart';
+
+void main() {
+  group('encode compact integers to hex', () {
+    test('Should return correct encoded data when value is zero', () {
+      final value = 0.toBigInt;
```
Comment (pastre) [inline packages/polkadart_scale_codec/test/test_types/codec_compact_test.dart:7]: nit: BigInt has an API for 0, 1 and 2 (chicken count lol)
```suggestion
      final value = BigInt.zero;
```

File: packages/polkadart_scale_codec/test/test_types/codec_compact_test.dart [hunk 5]
```diff
@@ -0,0 +1,104 @@
+import 'package:polkadart_scale_codec/polkadart_scale_codec.dart';
+import 'package:test/test.dart';
+
+void main() {
+  group('encode compact integers to hex', () {
```
Comment (pastre) [inline packages/polkadart_scale_codec/test/test_types/codec_compact_test.dart:5]: Test cases should be:
- [X] Negative int should throw
- [X] 0  should encode on first condition
- [ ] 63 should encode on first condition
- [ ] 64 should encode on second condition
- [ ] 2^14 - 1 should encode on second condition
- [X] 2^14 should encode on third condition
- [ ] 2^30 - 1 should encode on third condition
- [X] 2^30 should encode on last condition
- [ ] 2^536 - 1 should encode on last condition
- [ ] 2^536 should throw

I'm happy with an issue to refactor

Comment (pastre) [reply packages/polkadart_scale_codec/test/test_types/codec_compact_test.dart:5]: It's way more valuable to test the boundaries instead of a random value (like 69). If somebody was to change from 64 to 65, your tests would not fail
